### ISSUE TITLE
How would an extension to WAM work for a WP implementation?

### ISSUE THREAD

**Issue by iherman** [w3c/wpub#168]
In case the WP Manifest is defined as an extension to the WAM, and the deployment of WP (at first) is envisaged as some sort of an extension to the user agents, we have to make it sure that those extensions can access, via standard means, the data available in the Manifest.

**Comment by iherman** [comment]
(This description may lead to a new issue raised on the WAM spec.)

Background: the current [WAM spec (22 February version)](https://www.w3.org/TR/2018/WD-appmanifest-20180222/) describes a _processing_ model, whereby the user agent ["installs"](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#installation-process) a Web Applications, involving [processing the manifest](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#processing) (essentially, converting it into what is defined as an [IDL dictionary](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#webappmanifest-dictionary)), and registering service workers defined in the WAM instance. Once installation is done, an [event is raised](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#onappinstalled-attribute) (which can be caught, of course). Note that the installation procedure can be “controlled” (even interrupted or taken over by another script) via a separate [`beforeinstallprompt` event](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#onbeforeinstallprompt-attribute). 

The WAM spec also defines a possible [_extension point_](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#extensibility) to the WAM. This means that new members can be added to the WAM definition under certain constraints (e.g., no monkey patching existing definitions in the WAM). Extensions are required to define how to process new members, much the same way as the WAM currently does. These processing steps are, conceptually, “plugged in” the WAM manifest processing steps after all others terms are handled (see [step 14 in the processing steps](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#processing)).

In practice this means that if a user agent recognizes that the manifest contains extensions **_and_** it is also programmed to handled those extensions, then things are fine. (Although it is not clear _how_ a manifest signals the presence of extensions.) However, if the deployment approach is to provide some extra processing in the form of an extension/plugin/custom element/whatever then no hook seems to exist to provide for an additional processing to what the user agent does by default (ie, just handling the default WAM). Essentially, the only way is to provide an extension that ignores all the processing the user agent does, including processing the manifest, and take over processing altogether. This is certainly doable; however, it leads to an unnecessary duplication of processing, essentially duplicating the [manifest processing](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#processing).  

A possibility would be to define:

1. a callback possibility in [step 14 in the manifest processing steps](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#processing) to allow for an external function to complete the processing of the manifest data regarding the extensions
2. an extension of the Window object surfacing the [`WebAppManifest` dictionary instance](https://www.w3.org/TR/2018/WD-appmanifest-20180222/#webappmanifest-dictionary) so that scripts could re-use the data in the manifest after conversion.

That would allow extensions (e.g., an implementation of WAM affordances in a browser) to use a WP Manifest (if defined as a WAM extension), making use of the browser implementation of WAM.

**Comment by HadrienGardeur** [comment]
It's also worth noting that while the WAM itself allows extensions, there are no clear extension points for the `WebAppManifest` dictionary.

This means that the result of processing an extension to the WAM is a separate WebIDL dictionary, not an extension of the `WebAppManifest` dictionary.
That's aligned with [what we currently have in our draft](https://w3c.github.io/wpub/#webpublicationmanifest-dictionary).
